src/storage/src/dedup_pk_cell_based_row_deserializer.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::Deref;
+
+use risingwave_common::array::Row;
+use risingwave_common::catalog::OrderedColumnDesc;
+use risingwave_common::error::Result;
+use risingwave_common::types::{Datum, VirtualNode};
+use risingwave_common::util::ordered::OrderedRowDeserializer;
+
+use crate::cell_based_row_deserializer::{CellBasedRowDeserializer, ColumnDescMapping};
+
+/// Similar to [`CellBasedRowDeserializer`], but for dedup pk cell encoding.
+#[derive(Clone)]
+pub struct DedupPkCellBasedRowDeserializer<Desc: Deref<Target = ColumnDescMapping>> {
+    pk_deserializer: OrderedRowDeserializer,
+    inner: CellBasedRowDeserializer<Desc>,
+
+    // Maps pk fields with:
+    // 1. same value and memcomparable encoding,
+    // 2. corresponding row positions.
+    pk_to_row_mapping: Vec<Option<usize>>,
+}
+
+impl<Desc: Deref<Target = ColumnDescMapping>> DedupPkCellBasedRowDeserializer<Desc> {
+    /// Create a [`DedupPkCellBasedRowDeserializer`]
+    /// to decode cell based row with dedup pk encoding.
+    pub fn new(column_mapping: Desc, pk_descs: &[OrderedColumnDesc]) -> Self {
+        let (pk_data_types, pk_order_types) = pk_descs
+            .iter()
+            .map(|ordered_desc| {
+                (
+                    ordered_desc.column_desc.data_type.clone(),
+                    ordered_desc.order,
+                )
+            })
+            .unzip();
+        let pk_deserializer = OrderedRowDeserializer::new(pk_data_types, pk_order_types);
+
+        let pk_to_row_mapping = pk_descs
+            .iter()
+            .map(|d| {
+                let column_desc = &d.column_desc;
+                if column_desc.data_type.mem_cmp_eq_value_enc() {
+                    column_mapping
+                        .get(column_desc.column_id)
+                        .map(|(_, index)| index)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let inner = CellBasedRowDeserializer::new(column_mapping);
```

Comment:
Shall we remove deduped pks from column mapping?

Reply:
If we use original column mapping, when deserializing, the inner CellBasedRowDeserializer will leave placeholders for missing dedupped pk datums. Then we can just replace the placeholders with the dedupped pk datums.

If we remove deduped pks from column mapping, we can't do this.